File: problems/python/6. Zigzag Conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s

        res = [""] * numRows
        row = 0
        down = False

        for c in s:
            res[row] += c
            if row == 0 or row == numRows - 1:
                down = not down
            
            if down:
                row += 1
            else:
                row -= 1
        
        return "".join(res)

File: problems/Rust/6. Zigzag Conversion.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows == 1 || num_rows as usize >= s.len() {
            return s;
        }

        let n_rows = num_rows as usize;
        let mut res: Vec<String> = vec![String::new(); n_rows];
        let mut row: usize = 0;
        let mut down = false;

        for c in s.chars() {
            res[row].push(c);
            if row == 0 || row == n_rows - 1 {
                down = !down;
            }
            if down {
                row += 1;
            } else {
                row -= 1;
            }
        }
        res.join("")
    }
}